### ISSUE TITLE
Issue with Excel Wrapper for Coolprop for OS X Excel 2011

### ISSUE THREAD

**Issue by ragnar2015** [CoolProp/CoolProp#579]
I am trying to access CoolProp within Excel 2011 (OS X) using the instructions from here:
http://www.coolprop.org/coolprop/wrappers/Excel/

But Excel is not finding the dylib file evidently. Here are the steps that I followed...

I am using Mac OSX 10.6.8 and Excel 2011 version 14.4.5

I downloaded the file libCoolProp.dylib and placed it in a newly created directory:
/Users/nexus/lib      (nexus is the username)

Then I downloaded CoolProp.xlam and selected it as an Add-in at Excel/Tools. (It is in my download directory)

I copied the example call: PropsSI("T","P",101325,"Q",0,"Water") into an Excel cell and hit Enter.

Excel throws an error "The most recent error number is 53. Its message text is: File not found: libCoolProp.dylib

My default unix path is: PATH=/Users/nexus/Library/Python/2.7/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin

I did try adding the nexus/lib path to PATH but that did not fix the problem.

Also, I restarted the computer thinking that the new paths had to load.

So far nothing works.

Any ideas???

Don

**Comment by ibell** [comment]
Hmm.  Only thing I can think is that somehow you have either a 64-bit dylib and 32-bit Excel or vice versa.  Can you please give me the output of
``` file libCoolProp.dylib ``` 
on the shared library?  And the information from the about window in your Excel?  I know this definitely used to work, and I feel like it still should.  Should hopefully be easy to fix.

**Comment by ibell** [comment]
Hmm..  Just noticed you are on 10.6.8 - that's a rather old OS.  It is possible that Apple has changed where it looks for the dylib since 10.6.8.  Can you try putting it in the same directory with your sheet, just for fun?

**Comment by ragnar2015** [comment]
Hi Ian,

The output of "file libCoolProp.dylib" is one line:
libCoolProp.dylib: Mach-O dynamically linked shared library i386

I also placed the dylib file in the same directory where the Excel worksheet resides. Then I closed Excel and restarted it then opened the worksheet.  Same error 53. 

It is a head scratcher!

**Comment by ibell** [comment]
Yes, I agree.  That is weird. It seems like everything should be working
fine.  Let me try to fire up my OSX machine and see.

**Comment by ragnar2015** [comment]
I have to drop off for the night. I will check back in the morning and will help anyway that I can. Coolprop in Excel (2011) will be great when we can get over this speed bump.

**Comment by ibell** [comment]
I only worry that your OSX is too old, but let's see

**Comment by ibell** [comment]
Bad news.  It works just fine for me.  I am OSX 10.9.4, Excel 2011.

See the first accepted answer at this link for some more ideas:
https://social.msdn.microsoft.com/Forums/en-US/b060f291-0754-4e85-a7b9-e64259e6baad/vba-using-shared-library-office-2011-mac?forum=isvvba

I think you want to add ~/usr/lib to your DYLD_LIBRARY_PATH

See also
https://developer.apple.com/library/mac/#documentation/Darwin/Reference/Manpages/man1/dyld.1.html

**Comment by ragnar2015** [comment]
Hi Ian,

I have another machine with OSX 10.9.3 on it. I installed the dylib in ${HOME}/lib and CoolProp.xlam in a working directory. Same steps...  and it works like a charm!!!! Life is GOOD! Thanks.

My OSX 10.6.8 machine is a very stable work platform and it would be nice to get this working there. It looks like a path issue. Since we are in it, maybe others can benefit.

Per your suggestion, I searched google on adding  ~/usr/lib to the DYLD_LIBRARY_PATH. 

This website:
http://superuser.com/questions/282450/where-do-i-set-dyld-library-path-on-mac-os-x-and-is-it-a-good-idea 

warns as follows:
"One should never set export DYLD_LIBRARY_PATH on your system.
Shared library paths can be fixed using otool -L and install_name_tool."

Do you have specific instructions on how to safely add ~/usr/lib to the DYLD_LIBRARY_PATH ?

Thanks in advance!!

**Comment by ibell** [comment]
I'd say just try it - because you read someone aggressively saying don't do
something on the internet doesn't mean you shouldn't try - note that user
doesn't say WHY you shouldn't do that.   There have been times where I have
had to change it, for instance for the FORTRAN wrapper you see we have to
change it: http://www.coolprop.org/coolprop/wrappers/FORTRAN/index.html#run

So give that a try.  And please report back with the results.  If you could
make a pull request for this change, I'd be appreciative.

**Comment by jowr** [comment]
Have you tried making a symlink in one of the standard folders? I assume you have the library in your home folder: `sudo ln -s ~/usr/lib/libCoolProp.dylib /usr/local/lib`

**Comment by ragnar2015** [comment]
I executed this:    export DYLD_LIBRARY_PATH=${HOME}/lib
I verified using:    echo $DYLD_LIBRARY_PATH      
Result---> /Users/nexus/lib

I restarted Excel and opened a worksheet with the one cell containing   =PropsSI("T","P",101325,"Q",0,"Water")

Same 53 error.

**Comment by ibell** [comment]
But how can you be sure that Excel is picking up that DYLD_LIBRARY_PATH?  I
think you need to launch Excel from within that shell in which you set the
variable.

Or did you set it in your login script?

**Comment by ragnar2015** [comment]
I did not set it in a login script...I can try that after this.

Per your suggestion to open Excel in the same directory that I added the /lib path:
 To verify path is set within scope of this terminal session:
echo $DYLD_LIBRARY_PATH
/Users/nexus/lib       

open -a "/Applications/Microsoft Office 2011/Microsoft Excel.app"

Excel opened. I opened the worksheet and got the 53 error.

BTW what about the symlink method suggested by jowr?

**Comment by ibell** [comment]
This is really weird.  I wonder if they have changed the calling convention
or something like that.  I think that gives you the same error message.

**Comment by ibell** [comment]
Ah, or maybe it is not libCoolProp.dylib that is missing, rather one of its
dependencies.  Can you call "otool -L libCoolProp.dylib" on the dylib?

**Comment by ragnar2015** [comment]
Summarizing:  the Excel wrapper works in OSX Maverick (10.9) per your instructions here:
http://www.coolprop.org/coolprop/wrappers/Excel/
It is so simple and elegant and it DOES work! 

There is still an unknown issue with Excel 2011 in Mac OS 10.6 (Snow Leopard) whereby the dylib file is not accessible by Excel.

Ian, is this worth pursuing further?

**Comment by ibell** [comment]
Maybe, can you try the otool thing I recommended?  I suspect it is looking
for the wrong runtime library.

**Comment by ragnar2015** [comment]
Here you go (otool):

otool -L libCoolProp.dylib
libCoolProp.dylib:
        libCoolProp.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)

**Comment by ibell** [comment]
Hmm yes it links to libc++, which might not be a compatible version, let's
see if I can get it to statically link the c runtime, eliminating this
dependency.

**Comment by ibell** [comment]
Can you try with the nightly dylib I just built?  I'd be interested to see the output of otool -L libCoolProp.dylib as well.

**Comment by ibell** [comment]
In principle, I believe it should work.

**Comment by ragnar2015** [comment]
Hi Ian, I downloaded the nightly dylib from sourceforge that was posted "< 7 hours ago". Dropped it into my lib folder. Ran otool:

otool -L libCoolProp.dylib
libCoolProp.dylib:
        libCoolProp.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)

Then I closed Excel.
Verified that my lib folder was in the $DYLD_LIBRARY_PATH.
Opened Excel from the command line (same scope) and opened the worksheet with the PropsSI calls in it.

Same nemesis.....error 53 that says File not found: libCoolProp.dylib

**Comment by ibell** [comment]
And now, I think it should work (I hope).  For me:
```
Ians-Mac-mini:32bitso Ian$ file libCoolProp.dylib 
libCoolProp.dylib: Mach-O dynamically linked shared library i386
Ians-Mac-mini:32bitso Ian$ otool -L libCoolProp.dylib 
libCoolProp.dylib:
	/Users/Ian/Code/coolprop/build/32bitso/libCoolProp.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 60.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```

**Comment by ragnar2015** [comment]
Is there a new dylib? The last one on sourceforge is 17 hrs old as of 6pm USA east coast time.

**Comment by ibell** [comment]
There was a built  failure, give me 20 minutes, Ill let you know when its
up.

**Comment by ibell** [comment]
Ok, try now..

**Comment by ragnar2015** [comment]
Ian!  I downloaded the 3.2MB dylib file. Dropped it into my "lib" folder. Opened Excel and it WORKS!! No error 53!  The PropsSI calls are returning reasonable numbers. Next, I will exercise the coolprop functionality in Excel 2011. 

This is great work Ian, on not only this issue but CoopProp generally. It is extraordinarily useful. Thank you for your tenacity to resolving this issue (#579). 

Is there a donation jar somewhere?

BTW, will the "official" dylib file be OSX version agnostic or will there be separate versions?

**Comment by ibell** [comment]
The dylib that is up there will be essentially what will be released in 5.1
whenever that happens - one 32-bit dylib for 10.6 to 10.10.  I'm glad to
hear it finally works, and I'm surprised that no one ran into this problem
before.  Perhaps you are the only person that still uses 10.6 :)

Hmm, donation jar.  I should set something up, there are various means of
doing things like that now, like flattr and its ilk.

Finally closing this issue!
